# Data Cleaning

This notebook aims to clean the data for analysis. 

## Downloading data

- The data I'm using for this project is the [Street Tree List]('https://data.sfgov.org/City-Infrastructure/Street-Tree-List/tkzw-k3nq') from San Francisco Department of Public Works.  
- I downloaded the data on *Nov. 6, 2022*. At the time, the data set was last updated on Nov. 6, 2022 as well.
- I created a copy of the data set and named it `original_street_tree_list.csv`. I then put the data set in the `raw-data` folder.
- The dataset is in the repo already. You can also access it here: 


## Import the data

In [1]:
# import modules
import pandas as pd
import altair as alt

In [2]:
# read csv file
sf_trees_original = pd.read_csv('street_tree_list.csv')
sf_trees_original.head()

TreeID     qLegalStatus                                           qSpecies  \
0  217365  Section 806 (d)  Ceanothus 'Ray Hartman' :: California Lilac 'R...   
1   92771   DPW Maintained              Tristaniopsis laurina :: Swamp Myrtle   
2   23904   DPW Maintained     Prunus subhirtella 'Pendula' :: Weeping Cherry   
3   28646   DPW Maintained     Prunus subhirtella 'Pendula' :: Weeping Cherry   
4  229807   DPW Maintained                 Jacaranda mimosifolia :: Jacaranda   

           qAddress  SiteOrder                       qSiteInfo PlantType  \
0   707 Rockdale Dr        1.0  Sidewalk: Property side : Yard      Tree   
1   11X Blanken Ave        4.0    Sidewalk: Curb side : Cutout      Tree   
2  1600X Webster St        6.0                 Median : Cutout      Tree   
3  1600X Webster St        7.0                 Median : Cutout      Tree   
4    2560 Bryant St        1.0    Sidewalk: Curb side : Cutout      Tree   

  qCaretaker qCareAssistant               PlantDate  ...        XCoord  \
0    Private            NaN  10/14/2021 12:00:00 AM  ...  5.997488e+06   
1    Private            NaN  10/14/2021 12:00:00 AM  ...  6.011718e+06   
2        DPW            NaN                     NaN  ...  6.003596e+06   
3        DPW            NaN                     NaN  ...  6.003558e+06   
4    Private            NaN                     NaN  ...  6.009700e+06   

         YCoord   Latitude   Longitude  \
0  2.098235e+06  37.741209 -122.451285   
1  2.087394e+06  37.712247 -122.401320   
2  2.114195e+06  37.785380 -122.431304   
3  2.114375e+06  37.785872 -122.431449   
4  2.102427e+06  37.753411 -122.409355   

                                    Location  Fire Prevention Districts  \
0   (37.74120925101712, -122.45128526411095)                        9.0   
1  (37.712246915438215, -122.40132023435935)                       10.0   
2   (37.78537959802679, -122.43130418097743)                       13.0   
3   (37.78587163716589, -122.43144931782685)                       13.0   
4   (37.75341142310638, -122.40935530851043)                        2.0   

   Police Districts Supervisor Districts  Zip Codes  Neighborhoods (old)  
0               7.0                  4.0       59.0                 40.0  
1               3.0                  8.0      309.0                  1.0  
2               9.0                 11.0    29490.0                 13.0  
3               9.0                 11.0    29490.0                 13.0  
4               4.0                  7.0    28859.0                 19.0  

[5 rows x 23 columns]

In [3]:
sf_trees_original.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 196590 entries, 0 to 196589
Data columns (total 23 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   TreeID                     196590 non-null  int64  
 1   qLegalStatus               196533 non-null  object 
 2   qSpecies                   196590 non-null  object 
 3   qAddress                   195097 non-null  object 
 4   SiteOrder                  194796 non-null  float64
 5   qSiteInfo                  196590 non-null  object 
 6   PlantType                  196590 non-null  object 
 7   qCaretaker                 196590 non-null  object 
 8   qCareAssistant             24707 non-null   object 
 9   PlantDate                  70878 non-null   object 
 10  DBH                        153021 non-null  float64
 11  PlotSize                   146229 non-null  object 
 12  PermitNotes                53367 non-null   object 
 13  XCoord                     19

In [4]:
# Copy the original dataframe

sf_trees = sf_trees_original.copy()

## Check duplicates

In [5]:
# check if TreeID is unique
sf_trees['TreeID'].nunique()

196590

In [6]:
len(sf_trees)

196590

Seems like there's no duplicated `TreeID`. There are 196590 thousand trees planted in SF as of Nov.6, 2022. 

In [7]:
# make sure the length of dataframe matches the number of unique IDs

assert len(sf_trees) == sf_trees['TreeID'].nunique()

- There are 196590 TreeID, but only 70878 have a plant date. 
- `PlantDate` should also be a date Dtype.
- The `Zip Codes` are weird. There are some 2, 3 digit numbers.  

## Manage column names

In [8]:
# change all column names to lower case
sf_trees.columns = sf_trees.columns.str.lower()

In [9]:
sf_trees.columns = sf_trees.columns.str.replace(' ', '_')

In [10]:
sf_trees.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 196590 entries, 0 to 196589
Data columns (total 23 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   treeid                     196590 non-null  int64  
 1   qlegalstatus               196533 non-null  object 
 2   qspecies                   196590 non-null  object 
 3   qaddress                   195097 non-null  object 
 4   siteorder                  194796 non-null  float64
 5   qsiteinfo                  196590 non-null  object 
 6   planttype                  196590 non-null  object 
 7   qcaretaker                 196590 non-null  object 
 8   qcareassistant             24707 non-null   object 
 9   plantdate                  70878 non-null   object 
 10  dbh                        153021 non-null  float64
 11  plotsize                   146229 non-null  object 
 12  permitnotes                53367 non-null   object 
 13  xcoord                     19

## Convert Dtype

In [11]:
# convert the `PlantDate` Column to datetime

sf_trees['plantdate'] = pd.to_datetime(sf_trees['plantdate'])

In [12]:
sf_trees['treeid'] = sf_trees['treeid'].astype(object)

In [13]:
sf_trees.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 196590 entries, 0 to 196589
Data columns (total 23 columns):
 #   Column                     Non-Null Count   Dtype         
---  ------                     --------------   -----         
 0   treeid                     196590 non-null  object        
 1   qlegalstatus               196533 non-null  object        
 2   qspecies                   196590 non-null  object        
 3   qaddress                   195097 non-null  object        
 4   siteorder                  194796 non-null  float64       
 5   qsiteinfo                  196590 non-null  object        
 6   planttype                  196590 non-null  object        
 7   qcaretaker                 196590 non-null  object        
 8   qcareassistant             24707 non-null   object        
 9   plantdate                  70878 non-null   datetime64[ns]
 10  dbh                        153021 non-null  float64       
 11  plotsize                   146229 non-null  object  

In [14]:
sf_trees.head()

treeid     qlegalstatus                                           qspecies  \
0  217365  Section 806 (d)  Ceanothus 'Ray Hartman' :: California Lilac 'R...   
1   92771   DPW Maintained              Tristaniopsis laurina :: Swamp Myrtle   
2   23904   DPW Maintained     Prunus subhirtella 'Pendula' :: Weeping Cherry   
3   28646   DPW Maintained     Prunus subhirtella 'Pendula' :: Weeping Cherry   
4  229807   DPW Maintained                 Jacaranda mimosifolia :: Jacaranda   

           qaddress  siteorder                       qsiteinfo planttype  \
0   707 Rockdale Dr        1.0  Sidewalk: Property side : Yard      Tree   
1   11X Blanken Ave        4.0    Sidewalk: Curb side : Cutout      Tree   
2  1600X Webster St        6.0                 Median : Cutout      Tree   
3  1600X Webster St        7.0                 Median : Cutout      Tree   
4    2560 Bryant St        1.0    Sidewalk: Curb side : Cutout      Tree   

  qcaretaker qcareassistant  plantdate  ...        xcoord        ycoord  \
0    Private            NaN 2021-10-14  ...  5.997488e+06  2.098235e+06   
1    Private            NaN 2021-10-14  ...  6.011718e+06  2.087394e+06   
2        DPW            NaN        NaT  ...  6.003596e+06  2.114195e+06   
3        DPW            NaN        NaT  ...  6.003558e+06  2.114375e+06   
4    Private            NaN        NaT  ...  6.009700e+06  2.102427e+06   

    latitude   longitude                                   location  \
0  37.741209 -122.451285   (37.74120925101712, -122.45128526411095)   
1  37.712247 -122.401320  (37.712246915438215, -122.40132023435935)   
2  37.785380 -122.431304   (37.78537959802679, -122.43130418097743)   
3  37.785872 -122.431449   (37.78587163716589, -122.43144931782685)   
4  37.753411 -122.409355   (37.75341142310638, -122.40935530851043)   

   fire_prevention_districts  police_districts supervisor_districts  \
0                        9.0               7.0                  4.0   
1                       10.0               3.0                  8.0   
2                       13.0               9.0                 11.0   
3                       13.0               9.0                 11.0   
4                        2.0               4.0                  7.0   

   zip_codes  neighborhoods_(old)  
0       59.0                 40.0  
1      309.0                  1.0  
2    29490.0                 13.0  
3    29490.0                 13.0  
4    28859.0                 19.0  

[5 rows x 23 columns]

In [15]:
sf_trees['plantdate'].max()

Timestamp('2022-11-12 00:00:00')

In [16]:
sf_trees['plantdate'].min()

Timestamp('1955-09-19 00:00:00')

So the earliest tree plant date is in 1955. The tree is 67 years old now. 

## Sort dataframe

In [17]:
# sort by plantdate

sf_trees.sort_values(by=['plantdate'],ascending=True).head().reset_index(drop=True)

treeid    qlegalstatus                                qspecies  \
0  53719  Permitted Site                              Tree(s) ::   
1  30313  Permitted Site                              Tree(s) ::   
2  30312  Permitted Site                              Tree(s) ::   
3  30314  DPW Maintained  Pittosporum undulatum :: Victorian Box   
4  30316  Permitted Site  Acacia melanoxylon :: Blackwood Acacia   

             qaddress  siteorder                     qsiteinfo planttype  \
0     2963 Webster St        1.0  Sidewalk: Curb side : Cutout      Tree   
1     501 Arkansas St        3.0  Sidewalk: Curb side : Cutout      Tree   
2     501 Arkansas St        2.0  Sidewalk: Curb side : Cutout      Tree   
3     501 Arkansas St        1.0  Sidewalk: Curb side : Cutout      Tree   
4  1190 Sacramento St        6.0  Sidewalk: Curb side : Cutout      Tree   

  qcaretaker qcareassistant  plantdate  ...        xcoord        ycoord  \
0    Private            NaN 1955-09-19  ...  6.002896e+06  2.118758e+06   
1    Private            NaN 1955-10-20  ...  6.012995e+06  2.104700e+06   
2    Private            NaN 1955-10-20  ...  6.012995e+06  2.104700e+06   
3    Private            NaN 1955-10-20  ...  6.012998e+06  2.104677e+06   
4    Private            NaN 1955-10-24  ...  6.009098e+06  2.116730e+06   

    latitude   longitude                                   location  \
0  37.797869 -122.434054  (37.797868914156986, -122.43405362292302)   
1  37.759838 -122.398119   (37.75983788148773, -122.39811852764873)   
2  37.759838 -122.398119   (37.75983788148773, -122.39811852764873)   
3  37.759772 -122.398109   (37.75977229136789, -122.39810899635411)   
4  37.792650 -122.412449  (37.792650334753226, -122.41244910697198)   

   fire_prevention_districts  police_districts supervisor_districts  \
0                       13.0               9.0                  1.0   
1                       10.0               3.0                  8.0   
2                       10.0               3.0                  8.0   
3                       10.0               3.0                  8.0   
4                        3.0               1.0                 10.0   

   zip_codes  neighborhoods_(old)  
0       57.0                 17.0  
1    28856.0                 29.0  
2    28856.0                 29.0  
3    28856.0                 29.0  
4    28857.0                 21.0  

[5 rows x 23 columns]

For the data that has the `plantdate`, I want to create some new columns showing which year, month and day of the week those trees were planted.  

In [23]:
sf_trees['_day'] = sf_trees['plantdate'].dt.day_of_week.astype('Int64')
sf_trees['_month'] = sf_trees['plantdate'].dt.month.astype('Int64')
sf_trees['_year'] = sf_trees['plantdate'].dt.year.astype('Int64')


In [24]:
sf_trees.head(10)

treeid     qlegalstatus                                           qspecies  \
0  217365  Section 806 (d)  Ceanothus 'Ray Hartman' :: California Lilac 'R...   
1   92771   DPW Maintained              Tristaniopsis laurina :: Swamp Myrtle   
2   23904   DPW Maintained     Prunus subhirtella 'Pendula' :: Weeping Cherry   
3   28646   DPW Maintained     Prunus subhirtella 'Pendula' :: Weeping Cherry   
4  229807   DPW Maintained                 Jacaranda mimosifolia :: Jacaranda   
5  182857   DPW Maintained                  Afrocarpus gracilior :: Fern Pine   
6  266578   DPW Maintained          Magnolia grandiflora :: Southern Magnolia   
7  243397  Section 806 (d)                 Jacaranda mimosifolia :: Jacaranda   
8  178106   DPW Maintained     Photinia fraseri :: Photinia: Chinese photinia   
9  259361   DPW Maintained        Laurus nobilis :: Sweet Bay: Grecian Laurel   

             qaddress  siteorder                       qsiteinfo planttype  \
0     707 Rockdale Dr        1.0  Sidewalk: Property side : Yard      Tree   
1     11X Blanken Ave        4.0    Sidewalk: Curb side : Cutout      Tree   
2    1600X Webster St        6.0                 Median : Cutout      Tree   
3    1600X Webster St        7.0                 Median : Cutout      Tree   
4      2560 Bryant St        1.0    Sidewalk: Curb side : Cutout      Tree   
5       1831 27th Ave        1.0      Sidewalk: Curb side : Yard      Tree   
6      432 Holyoke St        2.0    Sidewalk: Curb side : Cutout      Tree   
7  253 Chattanooga St        1.0    Sidewalk: Curb side : Cutout      Tree   
8      475 Holyoke St        8.0    Sidewalk: Curb side : Cutout      Tree   
9      1699x Keith St        2.0    Sidewalk: Curb side : Cutout      Tree   

  qcaretaker qcareassistant  plantdate  ...   longitude  \
0    Private            NaN 2021-10-14  ... -122.451285   
1    Private            NaN 2021-10-14  ... -122.401320   
2        DPW            NaN        NaT  ... -122.431304   
3        DPW            NaN        NaT  ... -122.431449   
4    Private            NaN        NaT  ... -122.409355   
5    Private            NaN 2022-02-01  ... -122.485079   
6    Private            FUF        NaT  ... -122.408429   
7    Private            NaN 2021-10-15  ... -122.426337   
8    Private            NaN        NaT  ... -122.407999   
9        DPW            FUF 2021-10-26  ...         NaN   

                                    location fire_prevention_districts  \
0   (37.74120925101712, -122.45128526411095)                       9.0   
1  (37.712246915438215, -122.40132023435935)                      10.0   
2   (37.78537959802679, -122.43130418097743)                      13.0   
3   (37.78587163716589, -122.43144931782685)                      13.0   
4   (37.75341142310638, -122.40935530851043)                       2.0   
5   (37.75322320651839, -122.48507892882758)                       1.0   
6   (37.72583966005041, -122.40842894165918)                      10.0   
7   (37.75252386720633, -122.42633719012191)                       2.0   
8  (37.725395514253435, -122.40799924184167)                      10.0   
9                                        NaN                       NaN   

   police_districts  supervisor_districts  zip_codes  neighborhoods_(old)  \
0               7.0                   4.0       59.0                 40.0   
1               3.0                   8.0      309.0                  1.0   
2               9.0                  11.0    29490.0                 13.0   
3               9.0                  11.0    29490.0                 13.0   
4               4.0                   7.0    28859.0                 19.0   
5               8.0                   3.0       56.0                 35.0   
6               3.0                   7.0      309.0                 28.0   
7               4.0                   5.0    28862.0                 22.0   
8               3.0                   7.0      309.0                 28.0   
9               N

In [25]:
sf_trees.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 196590 entries, 0 to 196589
Data columns (total 26 columns):
 #   Column                     Non-Null Count   Dtype         
---  ------                     --------------   -----         
 0   treeid                     196590 non-null  object        
 1   qlegalstatus               196533 non-null  object        
 2   qspecies                   196590 non-null  object        
 3   qaddress                   195097 non-null  object        
 4   siteorder                  194796 non-null  float64       
 5   qsiteinfo                  196590 non-null  object        
 6   planttype                  196590 non-null  object        
 7   qcaretaker                 196590 non-null  object        
 8   qcareassistant             24707 non-null   object        
 9   plantdate                  70878 non-null   datetime64[ns]
 10  dbh                        153021 non-null  float64       
 11  plotsize                   146229 non-null  object  

In [ ]:
sf_trees

# Export the clean data

In [ ]:
sf_trees.to_csv('sf_trees_clean.csv', index=False)